fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/qrcode/QrScannerFragment.java

```diff
@@ -0,0 +1,198 @@
+    applyPermissionToView();
+  }
+
+  @Override
+  public void onDestroy() {
+    super.onDestroy();
+    if (barcodeScanner != null) {
+      barcodeScanner.close();
+    }
+  }
+
+  private ScanningAction getScanningAction() {
+    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
+      return requireArguments().getSerializable(SCANNING_KEY, ScanningAction.class);
+    } else {
+      return (ScanningAction) requireArguments().getSerializable(SCANNING_KEY);
+    }
+  }
+
+  private void setupNbScanned() {
+    viewModel
+        .getNbScanned()
+        .observe(getViewLifecycleOwner(), nb -> binding.scannedNumber.setText(String.valueOf(nb)));
+  }
+
+  private void setupAllowCameraButton() {
+    // Create request permission launcher which will ask for permission
+    ActivityResultLauncher<String> requestPermissionLauncher =
+        registerForActivityResult(
+            new ActivityResultContracts.RequestPermission(),
+            requireActivity().getActivityResultRegistry(),
+            isGranted -> applyPermissionToView() // This is the callback of the permission granter
+            );
+
+    // The button launch the build launcher when is it clicked
+    binding.allowCameraButton.setOnClickListener(
+        b -> requestPermissionLauncher.launch(Manifest.permission.CAMERA));
+  }
+
+  private void applyPermissionToView() {
+    if (checkSelfPermission(requireContext(), Manifest.permission.CAMERA) == PERMISSION_GRANTED) {
+      binding.cameraPermission.setVisibility(View.GONE);
+      binding.scannerInstructionText.setVisibility(View.VISIBLE);
+      binding.qrCodeSight.setVisibility(View.VISIBLE);
+      startCamera();
+    } else {
+      // the camera permission is not granted, make dedicated views visible
+      binding.cameraPermission.setVisibility(View.VISIBLE);
+      binding.scannerInstructionText.setVisibility(View.GONE);
+      binding.qrCodeSight.setVisibility(View.GONE);
+    }
+  }
+
+  private void startCamera() {
+    BarcodeScannerOptions options =
+        new BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_QR_CODE).build();
+    barcodeScanner = BarcodeScanning.getClient(options);
+    LifecycleCameraController cameraController = new LifecycleCameraController(requireContext());
+    cameraController.bindToLifecycle(this);
+    Executor executor = ContextCompat.getMainExecutor(requireContext());
+    cameraController.setImageAnalysisAnalyzer(
+        executor,
+        new MlKitAnalyzer(
+            Collections.singletonList(barcodeScanner),
+            COORDINATE_SYSTEM_VIEW_REFERENCED,
+            executor,
+            result -> {
+              List<Barcode> barcodes = result.getValue(barcodeScanner);
+              if (barcodes != null && !barcodes.isEmpty()) {
+                Log.d(TAG, "barcode raw value :" + barcodes.get(0).getRawValue());
+                onResult(barcodes.get(0));
+              }
+            }));
```

Comment:
I'm not sure I follow. Do you mean something similar as what you did in HBT ?

Reply:
But to be honest the manual add can be/is used to test the onResult method. So if the camera scanning part is mocked, I don't see what this would add in term of tested behaviour

Reply:
It's for karate 😄